cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/NamespaceSourceControlMetadataRefreshService.java

```diff
@@ -0,0 +1,314 @@
+    return new RepositoryTable(context);
+  }
+
+
+  // TODO(CDAP-21020): Refactor this function
+  private RepositoryMeta getRepositoryMeta(NamespaceId namespace)
+      throws RepositoryNotFoundException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      RepositoryTable table = getRepositoryTable(context);
+      RepositoryMeta repoMeta = table.get(namespace);
+      if (repoMeta == null) {
+        throw new RepositoryNotFoundException(namespace);
+      }
+
+      return repoMeta;
+    }, RepositoryNotFoundException.class);
+  }
+
+  // TODO(CDAP-21017): Optimize periodic refresh of source control metadata
+  @Override
+  public void runOneIteration() {
+    if (!running.compareAndSet(false, true)) {
+      return;
+    }
+    runRefreshService(false);
+  }
+
+  /**
+   * Triggers a manual refresh of source control metadata for the associated namespace.
+   *
+   * @param forced If true, forces the refresh by skipping the check whether the time elapsed since
+   *               the last refresh is within the buffer time.
+   */
+  public void triggerManualRefresh(boolean forced) {
+    if (!running.compareAndSet(false, true)) {
+      return;
+    }
+    executor().execute(() -> runRefreshService(forced));
+  }
+
+  private void runRefreshService(boolean forced) {
+    running.set(true);
+    Long refreshStartTime = System.currentTimeMillis();
+
+    try {
+
+      if (!namespaceAdmin.exists(namespaceId)) {
+        this.stop();
+      }
+
+      // Check if the time elapsed since the last refresh is within the buffer time.
+      // This is done  to prevent a high number of refresh jobs from being queued up
+      // unnecessarily.
+      if (!forced && System.currentTimeMillis() - lastRefreshTime.get() < bufferTime * 1000) {
+        return;
+      }
+
+      // Getting repository config for a specific namespace
+      RepositoryConfig repoConfig = getRepositoryMeta(namespaceId).getConfig();
+
+      // Listing the apps from remote repo
+      RepositoryAppsResponse repositoryAppsResponse = sourceControlOperationRunner.list(
+          new NamespaceRepository(namespaceId, repoConfig));
+
+      // Cleaning up the repo source control metadata table
+      HashSet<String> repoFileNames = repositoryAppsResponse.getApps().stream()
+          .map(RepositoryApp::getName)
+          .collect(Collectors.toCollection(HashSet::new));
+
+      cleanupRepoSourceControlMetaInBatches(namespaceId.getNamespace(), repoFileNames);
+
+      // Updating the namespace and repo source control metadata table
+      for (RepositoryApp repoApp : repositoryAppsResponse.getApps()) {
+        updateSourceControlMeta(
+            new ApplicationReference(namespaceId.getNamespace(), repoApp.getName()),
+            repoApp.getFileHash(),
+            refreshStartTime);
+      }
+
+      lastRefreshTime.set(refreshStartTime);
+
+    } catch (Exception e) {
+      LOG.error("Failed to refresh source control metadata for namespace "
```

Comment:
Use placeholder `{}` for logging instead of string concat. Also, seems like a warning than error.